server/src/main/java/org/elasticsearch/action/index/IndexRequest.java

```diff
@@ -591,20 +591,18 @@ public VersionType versionType() {
     }
 
     @Override
-    public void process() {
-        if ("".equals(id)) {
-            throw new IllegalArgumentException("if _id is specified it must not be empty");
-        }
+    public void process(IndexRouting indexRouting) {
+        indexRouting.process(this);
+    }
 
-        // generate id if not already provided
-        if (id == null) {
-            assert autoGeneratedTimestamp == UNSET_AUTO_GENERATED_TIMESTAMP : "timestamp has already been generated!";
-            assert ifSeqNo == UNASSIGNED_SEQ_NO;
-            assert ifPrimaryTerm == UNASSIGNED_PRIMARY_TERM;
-            autoGeneratedTimestamp = Math.max(0, System.currentTimeMillis()); // extra paranoia
-            String uid = UUIDs.base64UUID();
-            id(uid);
-        }
+    public void autoGenerateId() {
+        assert id == null;
+        assert autoGeneratedTimestamp == UNSET_AUTO_GENERATED_TIMESTAMP : "timestamp has already been generated!";
+        assert ifSeqNo == UNASSIGNED_SEQ_NO;
+        assert ifPrimaryTerm == UNASSIGNED_PRIMARY_TERM;
+        autoGeneratedTimestamp = Math.max(0, System.currentTimeMillis()); // extra paranoia
+        String uid = UUIDs.base64UUID();
+        id(uid);
```

Comment:
I found it quite convenient to add a method that can trigger the id auto-generation stuff. Mostly for tests.